File: messaging/services/whatsapp.py
import urllib.parse
from django.conf import settings


class WhatsAppMessageBuilder:
    """
    Builds WhatsApp deep links in a clean, testable way.
    """

    def __init__(self, phone_number: str):
        self.phone_number = phone_number

    def build_message(self, text: str) -> str:
        encoded_text = urllib.parse.quote(text)
        return f"https://wa.me/{self.phone_number}?text={encoded_text}"

    @staticmethod
    def appointment_message(data: dict) -> str:
        return (
            "Hello Dr Olaosebikan,\n\n"
            "I would like to book an appointment.\n\n"
            f"Name: {data.get('patient_name')}\n"
            f"Phone: {data.get('phone')}\n"
            f"Preferred Date: {data.get('preferred_date')}\n"
            f"Service: {data.get('service')}\n"
        )

    @staticmethod
    def general_message(data: dict) -> str:
        return (
            "Hello Dr Olaosebikan,\n\n"
            "I would like to make an enquiry.\n\n"
            f"Name: {data.get('patient_name')}\n"
            f"Phone: {data.get('phone')}\n"
            f"Message: {data.get('message')}\n"
        )

File: src/lib/whatsapp.ts
import { site } from "@/lib/content";

export type AppointmentMessageFields = {
  name?: string;
  phone?: string;
  email?: string;
  preferredDate?: string;
  preferredTime?: string;
  service?: string;
  message?: string;
};

export function whatsappUrl(message: string, phoneNumber = site.whatsappNumber) {
  return `https://wa.me/${phoneNumber}?text=${encodeURIComponent(message)}`;
}

export function appointmentMessage(fields: AppointmentMessageFields = {}) {
  const lines = [
    "Hello Dr Olaosebikan,",
    "",
    "I would like to book an appointment.",
    "",
    fields.name ? `Name: ${fields.name}` : "Name:",
    fields.phone ? `Phone: ${fields.phone}` : "Phone:",
    fields.email ? `Email: ${fields.email}` : "Email:",
    fields.preferredDate ? `Preferred Date: ${fields.preferredDate}` : "Preferred Date:",
    fields.preferredTime ? `Preferred Time: ${fields.preferredTime}` : "Preferred Time:",
    fields.service ? `Service/Complaint: ${fields.service}` : "Service/Complaint:",
    fields.message ? `Message/Symptoms: ${fields.message}` : "Message/Symptoms:",
  ];

  return lines.join("\n");
}

export function appointmentWhatsAppUrl(fields?: AppointmentMessageFields) {
  return whatsappUrl(appointmentMessage(fields));
}
